build: Add --disable-bip70 configure option

laanwj opened the pull request:
This patch adds a --disable-bip70 configure option that disables BIP70 payment request support in the wallet GUI. BIP70 support remains enabled by default. 

When disabled, this breaks the dependency of the GUI on OpenSSL and Protobuf.

(triggered by discussion on IRC today)

laanwj committed:
build: Add --disable-bip70 configure option

This patch adds a --disable-bip70 configure option that disables BIP70
payment request support. When disabled, this removes the dependency of
the GUI on OpenSSL and Protobuf.

TheBlueMatt commented:
Concept ACK generally, though I'm curious where you see this going longer-term - do we want to deprecate BIP 70 support (I think I'd be generally in favor of this, it seems to provide almost 0 utility which results in it being mostly unused, and even if it were used broadly, its unclear that it provides any real security benefit) or start shipping binaries without BIP 70 support or will this just be yet another build-time option we support (which I'd definitely use)?

MeshCollider commented:
Concept ACK after reading the discussion on IRC

promag commented:
Should add to travis matrix?

NicolasDorier commented:
For information, I plan to make BIP70 deprecated into NBitcoin. (By removing it from main lib, and moving it to separate package)

This led me to too much dependency issues, as well as cross implementation issues as you can't check correctly the signature of the payment request without serializing the payment request exactly as all other implementations does.  (Typically, my implementation works everywhere, except copay for reasons...)

As a library/service developer, I would love to see another simple standard, filling the same need, Json based, just using HTTPS for securing the communication between wallet and client.

laanwj commented:
> Concept ACK generally, though I'm curious where you see this going longer-term

My opinion on it is really divided. I like BIP70 in concept (automatic refund addresses, key expiration, allowing the wallet to directly authenticate vendors), but not the technical implementation, and also not the dependency burden it puts on bitcoin core.

Additionally it also seems the code is not maintained. No one is working on BIP70 support in bitcoin core.

So I'm ok with signalling that in the future we might be going to deprecate it, without any commitments - this would be the first step, to allow builders to disable it.

> For information, I plan to make BIP70 deprecated into NBitcoin. (By removing it from main lib, and moving it to separate package)

Seems to be better design anyhow. Here it's been peppered all over the GUI code :(

> Should add to travis matrix?

Agree that that's a good idea to warrant that this keeps working.

luke-jr commented on src/Makefile.qt.include:
```diff
@@ -161,6 +160,9 @@ QT_MOC_CPP = \
   qt/moc_walletmodel.cpp \
   qt/moc_walletview.cpp
 
+QT_MOC_CPP_BIP70 = \
```
IMO would be cleaner to just `QT_MOC_CPP +=` (and `BITCOIN_QT_CPP +=`) the actual files below, keeping all the stuff together.

laanwj replied:
I agree, but I just followed the flow already used in the makefile for the wallet, to not have to move large blocks around. A refactor like that could be done separately.

luke-jr commented on configure.ac:
```diff
@@ -1292,6 +1313,7 @@ echo "  with wallet   = $enable_wallet"
 echo "  with gui / qt = $bitcoin_enable_qt"
 if test x$bitcoin_enable_qt != xno; then
     echo "    qt version  = $bitcoin_qt_got_major_vers"
+    echo "    with bip70  = $enable_bip70"
```
Probably would be better to mention "(payment protocol)" here as well.

laanwj replied:
Yep

luke-jr commented on src/qt/bitcoin.cpp:
```diff
@@ -250,8 +252,10 @@ public Q_SLOTS:
     BitcoinGUI *window;
     QTimer *pollShutdownTimer;
-#ifdef ENABLE_WALLET
+#if defined(ENABLE_WALLET) && defined(ENABLE_BIP70)
```
Might be less messy to just leave the pointer here and not use it.

laanwj replied:
I don't think I agree. Accidentally using it would not generate a compile error in that case anymore.

promag replied:
Can't we just allow ENABLE_BIP70 if ENABLE_WALLET is true? In that case here `if defined(ENABLE_BIP70)` would be enough all over the place.

laanwj replied:
That'd make sense.

luke-jr commented on src/qt/bitcoin.cpp:
```diff
@@ -659,7 +667,7 @@ int main(int argc, char *argv[])
     initTranslations(qtTranslatorBase, qtTranslator, translatorBase, translator);
 
-#ifdef ENABLE_WALLET
+#if defined(ENABLE_WALLET) && defined(ENABLE_BIP70)
```
Won't this break opening non-BIP70 `bitcoin:` URIs?

laanwj replied:
Yes, as a by-effect this currently removes `bitcoin:` URL support too.
(it probably shouldn't)

Sjors replied:
That would break opening links from a browser or a mail client.

luke-jr commented on src/qt/coincontroldialog.cpp:
```diff
@@ -2,10 +2,15 @@
 #include "ui_coincontroldialog.h"
 
 #include "addresstablemodel.h"
+#include "base58.h"
```
These changes seem out of place here...?

laanwj replied:
These extra includes are necessary now that `paymentserver.h` doesn't include them anymore.

luke-jr commented on src/qt/guiutil.cpp:
```diff
@@ -9,6 +9,8 @@
 #include "walletmodel.h"
 
+#include "base58.h"
+#include "chainparams.h"
```
More out of place...

luke-jr commented on src/qt/test/compattests.cpp:
```diff
@@ -2,7 +2,13 @@
+
+#if defined(ENABLE_WALLET) && defined(ENABLE_BIP70)
 #include "paymentrequestplus.h" // this includes protobuf's port.h which defines its own bswap macos
+#endif
```
Do we need to get bswap macros from somewhere else?

laanwj replied:
No, we have our own bswap macros. The only reason this is tested is that there was a collision between our bswap macros and protobuf's. So commenting this out is harmless.

luke-jr commented on src/qt/test/wallettests.cpp:
```diff
@@ -1,5 +1,6 @@
 #include "wallettests.h"
 
+#include "base58.h"
```
Out of place

luke-jr reviewed (commented):
Concept ACK.

I suggest also indenting nested `#if`s with a single space.

Sjors commented:
"this breaks the dependency" -> "this *removes* the dependency"?

I know the goal is to get rid of OpenSLL, but what's with Protobuf? Just fewer dependencies, or is there a specific problem with it?

Breaking `bitcoin://1A1...Na&amount=0.001` would be bad; it's used by various services. It's often  embedded in a QR code, which isn't that useful on a desktop, but it can also be a link on a webpage or email too.

laanwj commented:
> I know the goal is to get rid of OpenSLL, but what's with Protobuf? Just fewer dependencies, or is there a specific problem with it?

Advantage of less dependencies is mainly: less stuff to build while cross-compiling, less attack surface (yet another parsing library), etc.

> Breaking bitcoin://1A1...Na&amount=0.001 would be bad; it's used by various services. It's often embedded in a QR code, which isn't that useful on a desktop, but it can also be a link on a webpage or email too.

I agree. It's `--disable-bip70` not `--disable-bip21`. @luke-jr already commented that, though.

laanwj committed:
qt: cleanup: Move BIP70 functions together in paymentserver

Reduces the number of separate `#ifdefs` spans.

laanwj commented:
I've restored BIP21 functionality, sifting through paymentserver.cpp/h to disable the parts relating to payment requests, instead of removing the whole file from the build.

luke-jr commented:
But `if (PaymentServer::ipcSendCommandLine())` is still #ifdef'd out...

laanwj commented:
> But if (PaymentServer::ipcSendCommandLine()) is still #ifdef'd out...

Whoops, missed that. Fixed.

sipa commented:
Concept ACK

jonasschnelli commented:
Nice!
Concept ACK. I could see this for 0.16 and a default to disable in 0.17.

> (my personal motivation is that I want to have the option to build the GUI without protobuf and without OpenSSL after the last remnants of OpenSSL use are removed from the rest of the code)

AFAIK OpenSSL (crypto) is still in use for the PRNG seeding (see currently closed #10299, waiting for new approach).
And as far as I can see there is a `RAND_event` in `winshutdownmonitor.cpp`.

laanwj commented:
> AFAIK OpenSSL (crypto) is still in use for the PRNG seeding (see currently closed #10299, waiting for new approach).

Agree. The rand_ stuff should be removed in one go in a separate PR, it's orthogonal to the changes here.

luke-jr commented on src/qt/bitcoin.cpp:
```diff
@@ -647,7 +651,7 @@ int main(int argc, char *argv[])
         QMessageBox::critical(0, QObject::tr(PACKAGE_NAME), QObject::tr("Error: %1").arg(e.what()));
         return EXIT_FAILURE;
     }
-#ifdef ENABLE_WALLET
```
We need this for BIP21 too.

Sjors commented:
I did a `make clean`, `./autogen.sh`, `./configure --disable-bip70` (which shows `with bip70  = no`) and `make deploy`.

I then tested BIP-21 using: `./Bitcoin-Qt.app/Contents/MacOS/Bitcoin-Qt bitcoin://1F1tAaz5x1HUXrCNLbtMDqcw6o5GNn4xqX?amount=0.1`. This worked, but it did throw a warning (?):

> QObject::connect: No such signal PaymentServer::receivedPaymentACK(QString) in qt/paymentserver.cpp:234

I think you need another `#ifdef` there.

laanwj commented:
> QObject::connect: No such signal PaymentServer::receivedPaymentACK(QString) in qt/paymentserver.cpp:234

Thanks, added, also rebased and squashed.

Sjors commented:
Warning is gone (also without the `//`).

fanquake commented:
Concept ACK. Needs another rebase.

laanwj commented:
Rebased for the #include change. Might want to re-review include changes.

laanwj commented:
Going to close this for now, [bitpay going to require BIP70](https://blog.bitpay.com/payment-protocol/) kind of messes up our plans to deprecate BIP70, and I'm not sure this option is worth carrying if that's not the long-term goal.

Sjors commented:
In that case, does it make sense to remove the OpenSSL dependency from the BIP70 implementation in some other way?

TheBlueMatt commented:
It is my understanding that there are enough other wallets that do not implement BIP 70 that it probably makes sense to just target deprecating it. Either way, I think we should *at least* be targeting moving towards a world in which the bitcoin URI which generates a payment request includes a signature over the payment request, because the use of https/TLS alone to authenticate the payment request probably makes it end up with significantly worse overall security than simply not using BIP 70.

NicolasDorier commented:
I disagree. This is just reinventing the wheel. 

If you want to separate the payment processor from the merchant, just make so the payment processor give an https link to the merchant, and show the merchant https certificate in the wallet transaction confirmation screen.

Adding your own protocol for signing a payment request is a bad idea.

TheBlueMatt commented:
@NicolasDorier absolutely not...If you are purchasing from a merchant, then you're already interacting with their website and *it* is what gave you the payment link, we should be using that as a root-of-trust where at all possible, not using *both* that *and* existing SSL CAs (after all, no user is going to verify that the merchant XYZ is actually using https://co1nbase.com/ as their payment processor, or that the 1 should really be an i and they're getting mitm'd).

NicolasDorier commented:
I am not sure I understand.

> I think we should at least be targeting moving towards a world in which the bitcoin URI which generates a payment request includes a signature over the payment request (or a pubkey which will sign it)

Which I disagree, because the payment request already come from HTTPS link. HTTPS can be used to show who you are paying to on the mobile wallet. (no need of adding your own signature on top of that)

> ..If you are purchasing from a merchant, then you're already interacting with their website and it is what gave you the payment link, we should be using that as a root-of-trust where at all possible, not using both that and existing SSL CAs

Which I completely agree but which seems to be the contrary of your previous sentence.

TheBlueMatt commented:
@NicolasDorier My point is you start from a root of trust (the bitcoin: URI), which no one is going to do almost anything to cross-check (cause if it references co1nbase.com instead of coinbase.com, who cares? thats the link you got). We should thus be using that as *the* root of trust, instead of introducing a second one (CAs), where if someone has a cert for coinbase.com (either cause they broke a CA somehow, or because you have a bad CA installed due to your work/antivirus software), they can intercept a payment even *after* you got a good bitcoin: URI!

NicolasDorier commented:
If you get the signed payment request inside the `bitcoin:`, how does the signer know that the pubkey signing it is from Coinbase? Response to this question is about re-implementing PKI, which browsers does out of the box, or is there any other alternatives?

TheBlueMatt commented:
The PKI usage in BIP70 is rarely used to verify the merchant itself, but is instead some third-party payment services provider (eg coinbase/bitpay/etc). Expecting to know which payment services provider a given merchant is using is not realistic for users, so I'm skeptical the PKI usage here is of really any value. Note that probably a better solution than requiring a signature over the payment request in the URI is requiring a public key which signs the eventual payment request in the URI. This also resolves the issue of having to get payment providers to get valid SSL/TLS certs for their domain and then use them for a secondary purpose (signing payment requests), which is generally (rightfully) considered bad practice.

NicolasDorier commented:
> The PKI usage in BIP70 is rarely used to verify the merchant itself, but is instead some third-party payment services provider (eg coinbase/bitpay/etc). Expecting to know which payment services provider a given merchant is using is not realistic for users, so I'm skeptical the PKI usage here is of really any value

Yes I agree, PKI as used in BIP70 is useless. What I am arguing is that by successfully downloading the payment request over a HTTPS channel, you also verified PKI through whatever arbitrary trust store you use underneath. (OpenSSL, Windows trust store, or iOS trust store)
Most of HTTP libraries in all languages are already doing it out of the box, so there is no need to implement our own.

> a better solution than requiring a signature over the payment request in the URI is requiring a public key which signs the eventual payment request in the URI

This is better, but I you still have the problem of identity misbinding. How do you know that the public key which signs is the merchant/coinbase? how can you be sure about that without relying on a PKI stack? (SSL/TLS)

TheBlueMatt commented:
> Yes I agree, PKI as used in BIP70 is useless. What I am arguing is that by successfully downloading the payment request over a HTTPS channel, you also verified PKI through whatever arbitrary trust store you use underneath. (OpenSSL, Windows trust store, or iOS trust store)

So what? You haven't gained anything in doing so aside from validating that your attacker is capable of getting an SSL cert issued, which is a free and automated process these days. Not sure what your point is here.

> This is better, but you still have the problem of identity misbinding. How do you know that the public key which sign is coming from the merchant/coinbase? how can you be sure about the identity behind the key without relying on a PKI stack (SSL/TLS) or a web of trust?

My point is you dont get that via BIP70 either, and I dont see that changing at any point in the future. Still, you *do* get this (somewhat) from bitcoin: URIs, as you may expect a user to only click such a link on the expected website's checkout flow (no different from where they enter their credit card). That is a very weak guarantee, but it is infinitely better than what BIP70 provides.

NicolasDorier commented:
Ah ok I think I understand you now.

```
bitcoin://?r=https://paymentprovider/id=invoiceId&merchantPubKey=abc
```

You then remove the PKI verification at BIP70 level (which is, as you said and I agree, useless).
You replace is by a signature embedded in the payment request of this pubkey.

By doing so, you are sure that the payment uri and the payment request are at least bound together.

I was confgused because in general this bitcoin: uri is not generated by the merchant's website but by the payment processor. If both, the uri and the payment request is generated by the payment processor, this scheme is not very useful.

MarcoFalke commented:
Why was this closed again? I think having an option that is disabled by default makes it easier for people that compile from source without having to pull in a ton of dependencies and build mess. See e.g. #14273, which could simply be worked around by setting `--disable-bip70`.

Sjors commented:
@gmaxwell I use BitPay occasionally with Bitcoin Core. The payment goes through just fine, but you get an error popup afterwards.

fanquake commented:
Removed "Up for grabs", this was done in #14451.
